src/shared/listener-collection.js

```diff
@@ -5,6 +5,23 @@
  * @prop {(event: Event) => void} listener
  */
 
+/**
+ * Return the event type that a listener will receive.
+ *
+ * For example `EventType<HTMLElement, 'keydown'>` evaluates to `KeyboardEvent`.
+ *
+ * The event type is extracted from the target's `on${Type}` property (eg.
+ * `HTMLElement.onkeydown` here) If there is no such property, the type defaults
+ * to `Event`.
+ *
+ * @template {EventTarget} Target
+ * @template {string} Type
+ * @typedef {`on${Type}` extends keyof Target ?
+ *   Target[`on${Type}`] extends ((...args: any[]) => void)|null ?
+ *     Parameters<NonNullable<Target[`on${Type}`]>>[0]
+ *  : Event : Event} EventType
```

Comment:
In words, the logic is:

1. Does `Target` have a `on${Type}` property?
2. Does the `on${Type}` property have the type we expect (`EventListener|null`)?
3. If it does, return that type with the `null` removed
4. If not, fall back to `Event`

Reply:
I partially grok. Follow-up questions:

* Why would we expect `null` as a valid type for `on${Type}`?
* Can you help me understand the `Parameters<NonNullable<Target[`on${Type}`]>>[0]` part? How does this "return that type with the `null` removed"? This looks like it would return the first argument to the listener function at first glance.
* The nested ternary here is a bit unfortunate but not TOO hard to read
* Would it be helpful for readability to type the "listener function" and then reference it as `ListenerFunction|null`? It might help to quasi-document this.

I think I can "handle this" with the exception of not being able to grok the `Parameters<>` clause.

Reply:
> Why would we expect null as a valid type for on${Type}?

If you read eg. `HTMLElement.onkeydown` before a handler has been assigned, you'll get `null`. Also you can set it to `null` to remove the handler.

> This looks like it would return the first argument to the listener function at first glance.

It does return the first argument, or rather, the type of the first argument, to the listener function. For a `keydown` event for example, the listener would be `(event: KeyboardEvent) => void` and so the type of the first argument is `KeyboardEvent`.

> The nested ternary here is a bit unfortunate but not TOO hard to read

Yeah. I don't know of an equivalent for `&&` to combine two "if" conditions into one.